expr/expr.go

```diff
@@ -24,209 +25,237 @@ var ErrNoSuchFunction = errors.New("no such function")
+	switch operator {
+	case "AND":
+		return &And{lhs, rhs}, nil
+	case "OR":
+		return &Or{lhs, rhs}, nil
+	default:
+		return nil, fmt.Errorf("unknown logical operator: %s", operator)
+	}
+}
+
+func evalBool(e Evaluator, rec *zng.Record) (zng.Value, error) {
+	zv, err := e.Eval(rec)
+	if err != nil {
+		return zv, err
+	}
+	if zv.Type != zng.TypeBool {
+		err = ErrIncompatibleTypes
+	}
+	return zv, err
+}
+
+func (a *And) Eval(rec *zng.Record) (zng.Value, error) {
+	lhs, err := evalBool(a.lhs, rec)
+	if err != nil {
+		return lhs, err
+	}
+	if !zng.IsTrue(lhs.Bytes) {
+		return zng.False, nil
+	}
+	rhs, err := evalBool(a.rhs, rec)
+	if err != nil {
+		return lhs, err
+	}
+	if !zng.IsTrue(rhs.Bytes) {
+		return zng.False, nil
+	}
+	return zng.True, nil
+}
+
+func (o *Or) Eval(rec *zng.Record) (zng.Value, error) {
+	lhs, err := evalBool(o.lhs, rec)
+	if err != nil {
+		return lhs, err
+	}
+	if zng.IsTrue(lhs.Bytes) {
+		return zng.True, nil
+	}
+	rhs, err := evalBool(o.rhs, rec)
+	if err != nil {
+		return lhs, err
+	}
+	if zng.IsTrue(rhs.Bytes) {
+		return zng.True, nil
+	}
+	return zng.False, nil
+}
+
+type In struct {
+	elem      Evaluator
+	container Evaluator
+	vals      Coercion
+}
+
+func compileIn(elem, container Evaluator) (Evaluator, error) {
+	return &In{
+		elem:      elem,
+		container: container,
 	}, nil
 }
 
-func compileIn(lhsFunc, rhsFunc NativeEvaluator) (NativeEvaluator, error) {
-	return func(rec *zng.Record) (zngnative.Value, error) {
-		rhs, err := rhsFunc(rec)
-		if err != nil {
-			return zngnative.Value{}, err
+func (i *In) Eval(rec *zng.Record) (zng.Value, error) {
+	container, err := i.container.Eval(rec)
+	if err != nil {
+		return container, err
+	}
+	typ := zng.InnerType(container.Type)
+	if typ == nil {
+		return zng.Value{}, ErrNotContainer
+	}
+	elem, err := i.elem.Eval(rec)
+	if err != nil {
+		return elem, err
+	}
+	iter := zcode.Iter(container.Bytes)
+	for {
+		if iter.Done() {
+			return zng.False, nil
 		}
-
-		typ := zng.InnerType(rhs.Type)
-		if typ == nil {
-			return zngnative.Value{}, ErrNotContainer
+		zv, _, err := iter.Next()
+		if err != nil {
+			return zng.Value{}, err
 		}
-
-		lhs, err := lhsFunc(rec)
+		_, err = i.vals.coerce(elem, zng.Value{typ, zv})
 		if err != nil {
-			return zngnative.Value{}, err
+			return zng.Value{}, err
 		}
-
-		iter := zcode.Iter(rhs.Value.(zcode.Bytes))
-		for {
-			if iter.Done() {
-				return zngnative.Value{zng.TypeBool, false}, nil
-			}
-
-			zv, _, err := iter.Next()
-			if err != nil {
-				return zngnative.Value{}, err
-			}
-			v, err := zngnative.ToNativeValue(zng.Value{typ, zv})
-			if err != nil {
-				return zngnative.Value{}, err
-			}
-			found, err := compare(lhs, v)
-			if err != nil {
-				return zngnative.Value{}, err
-			}
-			if found {
-				return zngnative.Value{zng.TypeBool, true}, nil
-			}
+		if i.vals.equal() {
+			return zng.True, nil
 		}
-	}, nil
+	}
 }
 
 func floatToInt64(f float64) (int64, bool) {
```

Comment:
The `floatToInt64` and `floatToUint64` functions remaining in the file are unused & can be deleted.

Reply:
I think this is a relic zngnative and reflects a change that I made... that 1.5:int will work but previously failed with error.  I will take them out, but are we okay with this change?  It general, I like an approach where zql does things that work and are robust but when surprising things happen there should be an easy figure things out (e.g., such coercion events could work rather than fail and be debounced and sent to the warnings channel).
```
% cat t.tzng
#0:record[x:int32,f:float64]
0:[1;1.5;]
% oldzq -t "put z=f:int64" t.tzng
bad cast
#0:record[x:int32,f:float64]
0:[1;1.5;]
% zq -t "put z=f:int64" t.tzng
#0:record[x:int32,f:float64,z:int64]
0:[1;1.5;1;]
```